### DISCUSSION TITLE
💙 Go

### DISCUSSION THREAD

**Discussion by mayukoooo** [mayukoooo/mercari-build-training#3] [学び]
memo

**Comment by mayukoooo** [comment]
- `log.Fatal` は使ってはいけないらしい
  - os.Exitが呼ばれてプログラムが強制終了される（サーバープロセス全体が落ちてしまう）
  - プログラムが強制終了されて何がまずいかというと、**通常Goではdeferで後処理を予約するようなコードを多用しますが強制終了してしまうとこれらの後処理が呼ばれずリソースの適切な解放などが行われないまま終了してしまう**
  - 代替案
    -  書き捨てのコードなどでより手抜きしたいという場合は`log.Panic`
      -  defer が処理される
    - 別の関数にエラーハンドリングを切り出す

```go
func run() error {
    // 処理本体
}

func main() {
    err := run()
    if err != nil {
        log.Fatal(err)
    }
}
```

**Reply by mayukoooo** [reply]
https://zenn.dev/snowcrush/articles/21f28163e067cb

**Reply by task4233** [reply]
今回利用しているGoのWeb serverであるEcho Serverでは、Panicを使う代わりに `echo.NewHTTPError` で生成されたエラーで一元的にHTTPエラーハンドリングする方法が提唱されています。

ref: [Error Handling | Echo](https://echo.labstack.com/docs/error-handling)

**Reply by task4233** [reply]
また、Goではライブラリを実装する際に、Mayu-sanが言及していたFatal系の関数に加え、panicも基本的には利用しないことが推奨されています。代わりに、errorを返すことが推奨されています :+1:

> This is only an example but real library functions should avoid panic. If the problem can be masked or worked around, it's always better to let things continue to run rather than taking down the whole program. One possible counterexample is during initialization: if the library truly cannot set itself up, it might be reasonable to panic, so to speak.
ref: https://go.dev/doc/effective_go#panic

**Reply by mayukoooo** [reply]
関連（review comments）：
> must: log.Panic は内部的にpanic()を実行するので、使うのを避けてください(今回利用しているEchoというWebサーバは、panicしてもrecoverする機能があるのでサーバは終了しませんが) 🙏
ref: https://pkg.go.dev/log#Panic
>
>代わりに、ログを残したい場合は c.Logger()が使えます 👍
>
>[利用例](https://github.com/mercari-build/mercari-build-training/blob/main/go/app/main.go#L48)
[利用できるメソッド群](https://pkg.go.dev/github.com/labstack/echo/v4#Logger)

https://github.com/mayukoooo/mercari-build-training/pull/4#discussion_r1493792473


> fyi: ちなみに、Echo serverでStatus Codeを返却する例は公式ページで提示されているので、参考にしつつ適切なStatus Codeを設定すると良いと思います！
ref: [Error Handling | Echo](https://echo.labstack.com/docs/error-handling)

https://github.com/mayukoooo/mercari-build-training/pull/4#discussion_r1493793476

**Comment by mayukoooo** [comment]
- Unmarshal
  - JSON形式で受け取った値を指定した構造体に格納することができる
  - 第1引数：JSON形式のデータ
  - 第2引数：格納したい構造体

**Reply by mayukoooo** [reply]
- json.Marshal
 - Item型のオブジェクトをJSON形式のバイト列に変換する

**Reply by task4233** [reply]
ref: 
- [func Unmarshal | encoding/json](https://pkg.go.dev/encoding/json#Unmarshal)
- [func Marshal | encoding/json](https://pkg.go.dev/encoding/json#Marshal)

**Comment by mayukoooo** [comment]
- c.FormValue("name")
  - HTTPリクエストから"name"という名前のフォームデータを取得するための関数
  - echo.Contextオブジェクトのメソッドとして提供されている
  - フォームデータの値を文字列として返す
  - もしリクエストに"name"という名前のフォームデータが存在しない場合は、空の文字列を返す

**Comment by mayukoooo** [comment]
- c.Logger().Infof("Receive item: %s", name)
  - ログに"Receive item: <nameの値>"というメッセージを出力するためのコード
  - %s
    - name変数の値に置き換えられる
  - c.Logger()
    - Echoフレームワークのロギング機能を提供
  - Infofメソッド
    - 情報レベルのログを出力する

**Comment by mayukoooo** [comment]
- fmt.Sprintf
  - 指定されたフォーマットに従って文字列を生成するための関数
  - フォーマット文字列とそれに続く任意の数の引数を受け取り、フォーマット文字列内のフォーマット指定子（例えば%sや%dなど）を引数の値に置き換えて、その結果を新しい文字列として返す
```go
name := "world"
message := fmt.Sprintf("Hello, %s!", name)
```

**Comment by mayukoooo** [comment]
- json.Marshal
 - Item型のオブジェクトをJSON形式のバイト列に変換する

**Comment by mayukoooo** [comment]
- os.WriteFile
  - 第一引数：指定されたファイル名
  - 第二引数：ファイルに書き込むデータ
  - 第三引数：ファイルのパーミッションを指定
  - ファイルが存在しない場合は新しく作成してくれる
  - 存在する場合は内容を上書きする
  - 0644
    - オーナーだけ書き込みできる
    - https://www.profuture.co.jp/mk/column/41710#:~:text=%E3%80%8C644%E3%80%8D%E3%81%AF%E3%80%81%E3%82%AA%E3%83%BC%E3%83%8A%E3%83%BC%E3%81%AF,%E3%81%A7%E3%81%8D%E3%81%A6%E3%81%97%E3%81%BE%E3%81%86%E3%81%AE%E3%81%A7%E3%81%99%E3%80%82

**Reply by mayukoooo** [reply]
使用例
```go
	if err := os.WriteFile(itemsJson, updatedData, 0644); err != nil {
		log.Panic(err)
	}
```

**Comment by mayukoooo** [comment]
- ファイルはopenしたらcloseする
  - open：ファイルへのアクセスを開始する
    - openしてから、読み取りやファイルの内容の処理等を行う
  - close：ファイル操作が完了した後にファイルを適切に閉じる
    - リソースリークを防ぎ、システムリソースを適切に管理するために重要 
  - defer：関数が終了する直前にファイルを閉じることを保証する。これにより、ファイル操作が完了した後にリソースが適切に解放される。
```go
f, err := os.Open("text.txt")
if err != nil {
    fmt.Println("cannot open the file")
}
defer f.Close()
```

**Comment by mayukoooo** [comment]
- ハッシュ化
```go
hashedValue := sha256.New() // SHA256でハッシュ化
```

**Comment by mayukoooo** [comment]
- io.Copy
  - 実行成功した場合、コピーされたバイト数と、発生した可能性のあるエラーを返す
```go
io.Copy(コピー先, コピー元)
```
- 例えば第一引数にハッシュ関数、第二引数に画像ファイルを指定したら、画像の内容に基づいたハッシュ値が生成される
```go
	if _, err := io.Copy(sha256.New(), src); err != nil {
		return err
	}
```

**Reply by mayukoooo** [reply]
- hoge.Sum(nil)
  - ハッシュ関数によって計算されたハッシュ値を取得するために使用する
  - nilを引数として渡すことで、これまでにハッシュ関数に書き込まれたデータに基づいて最終的なハッシュ値を計算し、その値をバイトスライスとして返す
```go
	hash := sha256.New()
	if _, err := io.Copy(hash, src); err != nil {
		return err
	}
	hashed := fmt.Sprintf("%x", hash.Sum(nil)) + ".jpg"
```

**Reply by mayukoooo** [reply]
Goでの "+" operatorを用いた文字列連結は低速なので、".jpg"の部分もfmt.Sprintfへ

**Comment by mayukoooo** [comment]
- os.Create
  - 新しいファイルを作成する
- path.Join
  - 第一引数：指定されたディレクトリ
  - 第二引数：保存したいファイル

**Comment by mayukoooo** [comment]
- range
  - sliceやmapなどのコレクションを反復処理するために使用する

**Reply by mayukoooo** [reply]
goのデータ型
- slice:
  - 可変長の配列を表す
  - 同じ型の要素を順序付けて格納し、そのサイズは実行時に変更可能
- map:
  - キーと値のペアのコレクションで、キーに対して値を格納・検索するために使用する
  - キーは一意でなければいけない

```go
for index, value := range slice {
    // indexはスライスのインデックス、valueはそのインデックスの値
}

for key, value := range map {
    // keyはマップのキー、valueはそのキーに対応する値
}
```

**Reply by mayukoooo** [reply]
使用例：idが一致するものを探す時
```go
	for _, item := range itemList.Items {
		if item.ID == id {
			return c.JSON(http.StatusOK, item)
		}
	}
```

**Comment by mayukoooo** [comment]
- `json.Unmarshal` vs `json.NewDecoder`

> この例で言うと json.NewDecoder() を使う方が良いです！
> 理由は[この記事](https://qiita.com/Coolucky/items/44f2bc6e32ca8e9baa96)に書かれている通りで、 json.Unmarshal() のためにはデータを途中で持たなくてはならないためです
> (一応 json.NewDecoder() も内部的にデータを持っていますが、データの持ち方が効率化されているので、こちらの方が良いです)
> 字が汚くて申し訳ないのですが、イメージは添付した画像の通りです :+1:

![IMG_0186](https://github.com/mayukoooo/mercari-build-training/assets/98243891/56069fd0-2c5a-45bb-9a2e-8012e80174c3)

**Reply by mayukoooo** [reply]
使用例
```go
	var itemList ItemList
	newData := bytes.NewReader(data)
	if error := json.NewDecoder(newData).Decode(&itemList); error != nil {
		getErrorStatus(c, "Failed to newReader items.json")
	}
```

**Reply by mayukoooo** [reply]
- bytes.NewReader(data)
  - dataというバイトスライスから新しい*Readerを作成している
  - dataを読み込むための読み取りインターフェースが提供され、json.NewDecoderに渡すことができる

- json.NewDecoder
  - io.Readerインターフェースを実装する任意のオブジェクトからデコードを行う
  - bytes.NewReaderを使用してdataバイトスライスをio.Readerとして扱えるようにしている

**Comment by mayukoooo** [comment]
- DBから必要な値だけを取った方が安全
```go
rows, err := db.Query("SELECT * FROM items")
```
↓
```go
rows, err := db.Query("SELECT name, category, image_name FROM items")
```

**Comment by mayukoooo** [comment]
- 変数を初期化するか否か
```go
var items Items
```
- Items型のitemsを宣言
- 明示的に初期化は行なっていない
- goでは変数を宣言すると自動的にその型のゼロ値で初期化される
  - -> 型のゼロ値で変数を初期化している

```go
items := Items{Items: []*Item{}}
```
- Items型の変数itemsを宣言し、同時に初期化している
  - Items構造体のItemsフィールドを空のItemポインタのスライスで初期化
  - -> これにより、itemsは空のスライスを持つItemsインスタンスとして初期化される

**Comment by mayukoooo** [comment]
- スライス
  - Go言語における動的な配列の一種
  - 固定長の配列と異なり、スライスはサイズが動的に変更可能で、要素の追加や削除が容易
  - スライスは、配列の一部または全部を参照するための構造体
    - ポインタ: スライスの最初の要素を指す
    - 長さ: スライスに含まれる要素の数
    - 容量: スライスの基となる配列が保持できる要素の最大数

**Comment by mayukoooo** [comment]
- Itemのスライスが値として保持されている
  - -> その要素が直接スライス内に格納される
```go
type Items struct {
	Items []Item `json:"items"`
}
```

- Itemの**ポインタ**のスライスが保持されている
  - -> 要素への参照（アドレス）がスライス内に格納される
    - -> ポインタを使用すると、メモリの使用量を節約し、要素の共有や変更が容易になる
    - -> 扱いが少し複雑になる場合もあり
```go
type Items struct {
	Items []*Item `json:"items"`
}
```

**Reply by mayukoooo** [reply]
https://qiita.com/kujirahand/items/d6952b8b1f4793564af3

**Reply by mayukoooo** [reply]
ポインタ
https://qiita.com/HiromuMasuda0228/items/65b9a593275f769f6b69#%E3%83%9D%E3%82%A4%E3%83%B3%E3%82%BF

**Comment by mayukoooo** [comment]
エラーハンドリング

**Reply by mayukoooo** [reply]
https://qiita.com/immrshc/items/13199f420ebaf0f0c37c

**Reply by mayukoooo** [reply]
https://go.dev/blog/error-handling-and-go

**Reply by mayukoooo** [reply]
https://zenn.dev/spiegel/books/error-handling-in-golang/viewer/evaluations

**Reply by mayukoooo** [reply]
```go
func parseError(c echo.Context, message string, err error) {
        // 自分が設定したエラーメッセージをコンソールに表示する
        // 簡易的にどこに問題がありそうかがわかる
	res := Response{Message: message}
	c.JSON(http.StatusInternalServerError, res);

        // 具体的にどこに問題があるのかコンソールに出力できる
	c.Logger().Error(err);
}
```

**Comment by mayukoooo** [comment]
- strconv.Atoi関数
  - 文字列をint型に変換し、変換できた整数とエラー（変換できなかった場合）を返す
  - エラーがnilでない場合、変換に失敗したことを意味する
